### ISSUE TITLE
coalesce function output differs from documentation

### ISSUE THREAD

**Issue by rachlmac** [hashicorp/terraform#20982]
### Terraform Version

```
Terraform v0.12.0-beta1
```

### Terraform Configuration Files

```hcl
output "coalesce_a" {
  value = coalesce("a", "b")
	}

output "coalesce_b" {
  value = coalesce("", "b")
	}

```

### Expected Behavior

Outputs:

coalesce_a = a
coalesce_b = b

### Actual Behavior

Outputs:

coalesce_a = a
coalesce_b =

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Additional Context

I originally discovered this when using terraform console, and then verified it by actually executing terraform with the same values. I'm not sure if this is the intended behavior of coalesce (it is not what I expected, based off of the documentation and my general understanding of the use of coalesce) and therefore the documentation may need to be updated, or if this is not the intended behavior and is then a bug with coalesce.
### References

Thank you for your time!

**Comment by apparentlymart** [comment]
Hi @rachlmac! Thanks for reporting this.

Indeed, this looks like a bug. Specifically, I think we misled ourselves due to the function name into thinking that the `coalesce` function provided by an underlying library in Terraform's configuration language interpreter was a suitable implementation for Terraform's `coalesce`, but the behavior is actually different.

The currently-used `coalesce` has the behavior of selecting the first _non-null_ element from a sequence of values of any type. _Terraform's_ coalesce, on the other hand, selects the first _non-empty_ string from a sequence of _strings_.

To resolve this, we should copy the upstream [`CoalesceFunc`](https://github.com/zclconf/go-cty/blob/3ccef1ae92b0ac73b92d86f274f0ff781e2d9b3b/cty/function/stdlib/general.go#L57) into Terraform's own functions package and adapt it into what Terraform's documentation suggests, since that will ensure compatibility with prior versions of Terraform. We can adapt the unit tests from the Terraform v0.11 `coalesce` function to ensure that we're hitting all the same cases.

I expect that a function that returns the first non-null value in a sequence will also be useful now that the Terraform language _has_ the concept of a null value (it didn't prior to v0.12), compatibility must be the priority here. While fixing this, we could also experiment to see if we can add that functionality to `coalesce` in a backward-compatible way, or whether it'll be necessary/better to add a new function for that instead (which would be unfortunate, since the name `coalesce` _sounds_ like it deals in nulls from experience with other software). I _think_ it is possible, as follows:

* The v0.11 `coalesce` required all of its arguments to be strings anyway, and so we can safely accept any other type without breaking compatibility because no other type has a value that would convert to an empty string and thus be treated as empty by the old implementation.
* We can expand the rules for this function to skip over any null _or_ empty string values safely, because the idea of null values didn't exist at all in v0.11 and so nobody can be depending on that behavior.

It is a little quirky/unfortunate that in this function alone we'd be considering an empty string to be equivalent to a null value, but in the grand scheme of things I think that's a pragmatic compromise to get us the `coalesce` function we want while remaining compatible with the `coalesce` function we had.

Thanks again for reporting this, @rachlmac! We'll get this fixed up before the final v0.12.0 release.

**Comment by rachlmac** [comment]
Hey @apparentlymart thanks for your response, it has been super useful in allowing me to find a work around for the present. I don't know if it is necessary to give this additional context, but I figure more information is better than less:

Filling out this bug tickled a memory of an issue I ran into doing a 0.12 upgrade that I initially brushed off as user error or growing pains of an (otherwise excellent) early release.
I was upgrading a configuration that created an aws_instance resource, that had a section containing a remote-exec. Post running the upgrade command, it was re-written to:
```
provisioner "remote-exec" {
connection {
host                = coalesce(self.public_ip, self.private_ip)
type                = "ssh"
(...etc...)
}
```
This inevitably failed with `Error: timeout - last error: ssh: rejected...`, as the resources I provisioned only had private_ips and, at least from what I understand now about the 0.12 implementation of coalesce, host was being populated with an empty string from this.

As a sanity check, I just had this config file output 
```
output "coalesce_ip" {
    value = coalesce(aws_instance.node[0].public_ip, aws_instance.node[0].private_ip)
      }
```
which resulted in an empty string output.

**Comment by mildwonkey** [comment]
Hi @rachlmac! 
I just wanted to let you know that this bug is fixed in #21002, but I am leaving this issue open to remind us to make a note about in the upgrade guide.
Thank you again for reporting this!

**Comment by apparentlymart** [comment]
In order to avoid the upgrade guide growing unreasonably large, we've decided for now to not call this particular change out explicitly, since we expect the slight changes of behavior in `coalesce` (ignoring null values, and producing an error rather than an empty string if nothing is selected) will not cause problems for most existing configurations, and for any that _were_ relying on the behavior of returning an empty string when nothing matches, the new error message will serve as the prompt to update it.

We'll keep an eye out during the remaining prereleases to see if that assumption is proven incorrect by others running into it, and revisit this decision if so. For now though, I think the work here is done and so I'm going to close this out.

Thanks again for reporting this, @rachlmac!
